# Q1

In [1]:
import pandas as pd

a = pd.Series({'jan': 100, 'feb': 200, 'mar': 300})
b = pd.Series({'feb': 10, 'mar': 20, 'apr': 40})

result = a + b

print(result)

apr      NaN
feb    210.0
jan      NaN
mar    320.0
dtype: float64


# Prediction: jan = NaN, feb = 210.0, mar = 320.0, and apr = NaN.

# Actual result: The prediction matches the output.

# Explanation: Pandas aligns Series by their index labels before performing arithmetic. 
# The result therefore contains the union of the two indexes: jan, feb, mar, and apr. Values are added where both Series have the same label, 
# while labels present in only one Series produce NaN.

In [5]:
import pandas as pd

df = pd.read_excel(r"C:\Users\Elitebook 1030 G2\Downloads\cAlwUp3sSzSjtoyrfBXL_saas (1).xlsx", dtype=str)

print(df.head())

  account_id    company            signup_ts        country seats   mrr  \
0     A-1001  Acme Corp  2023-01-05 09:00:00             US    25  2500   
1     A-1002   Beta LLC  2023-02-14 00:00:00             US     5   450   
2     A-1001  Acme Corp  2023-01-05 09:00:00             US    25  2500   
3     A-1003     Cirrus           1675209600             GB     0     0   
4     A-1004  Delta Inc  2023-03-20 00:00:00  United States    12  1188   

         plan          last_active churned  
0  enterprise  2024-03-01 00:00:00   False  
1         pro  2024-03-02 00:00:00   False  
2  enterprise  2024-03-01 00:00:00   False  
3         pro  2024-02-28 00:00:00   False  
4         pro  2024-03-01 00:00:00      No  


# Q2

In [6]:
print(type(df['mrr']))
print(type(df[['mrr']]))
print(type(df.loc[:, 'mrr']))
print(type(df.loc[:, ['mrr', 'seats']]))

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>


df['mrr'] returns a Series because a single column label is selected. df[['mrr']] returns a DataFrame because a list of columns is selected. 
Similarly, .loc[:, 'mrr'] returns a Series, while .loc[:, ['mrr', 'seats']] returns a DataFrame. 
This matters because operations expecting a two-dimensional DataFrame, such as selecting multiple columns or preserving DataFrame structure, behave differently from operations on a one-dimensional Series.

# Q3

In [7]:
df[['company', 'seats', 'mrr']]

,company,seats,mrr
0,Acme Corp,25,2500
1,Beta LLC,5,450
2,Acme Corp,25,2500
3,Cirrus,0,0
4,Delta Inc,12,1188
5,Echo,8,-99
6,Foxtrot,3,270
7,Gamma,15,1800
8,Helio,NaN,540
9,Iris,10,900


The four direct selections are df[['company', 'seats', 'mrr']], df.loc[:, ['company', 'seats', 'mrr']], df.iloc[:, [1, 4, 5]], and df.filter(items=['company', 'seats', 'mrr']). 
The positional .iloc approach is fragile because it depends on column positions, and a .loc range depends on the relative ordering of columns.
I would use explicit column-name selection in production because the requested fields remain clear even if unrelated columns are reordered.

In [ ]:
Q4

In [10]:
df.loc[:, 'company':'mrr']

,company,signup_ts,country,seats,mrr
0,Acme Corp,2023-01-05 09:00:00,US,25,2500
1,Beta LLC,2023-02-14 00:00:00,US,5,450
2,Acme Corp,2023-01-05 09:00:00,US,25,2500
3,Cirrus,1675209600,GB,0,0
4,Delta Inc,2023-03-20 00:00:00,United States,12,1188
5,Echo,2023-04-01 00:00:00,US,8,-99
6,Foxtrot,2023-05-11 00:00:00,us,3,270
7,Gamma,2023-06-30 00:00:00,DE,15,1800
8,Helio,2023-07-22 00:00:00,US,NaN,540
9,Iris,2023-08-08 00:00:00,US,10,900


In [11]:
df.iloc[:, 1:5]


,company,signup_ts,country,seats
0,Acme Corp,2023-01-05 09:00:00,US,25
1,Beta LLC,2023-02-14 00:00:00,US,5
2,Acme Corp,2023-01-05 09:00:00,US,25
3,Cirrus,1675209600,GB,0
4,Delta Inc,2023-03-20 00:00:00,United States,12
5,Echo,2023-04-01 00:00:00,US,8
6,Foxtrot,2023-05-11 00:00:00,us,3
7,Gamma,2023-06-30 00:00:00,DE,15
8,Helio,2023-07-22 00:00:00,US,NaN
9,Iris,2023-08-08 00:00:00,US,10


df.loc[:, 'company':'mrr'] returns five columns: company, signup_ts, country, seats, and mrr. .loc label slicing includes both endpoint labels. 
df.iloc[:, 1:5] returns four columns: company, signup_ts, country, and seats, because .iloc uses positional Python-style slicing where the ending position is excluded.

# Q5. The count is the question

In [12]:
print("Shape of dataset:")
print(df.shape)

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

Shape of dataset:
(17, 9)

Column names:
['account_id', 'company', 'signup_ts', 'country', 'seats', 'mrr', 'plan', 'last_active', 'churned']

Data types:
account_id     object
company        object
signup_ts      object
country        object
seats          object
mrr            object
plan           object
last_active    object
churned        object
dtype: object


In [13]:
print("Missing values in each column:")
print(df.isna().sum())

Missing values in each column:
account_id     0
company        1
signup_ts      0
country        0
seats          1
mrr            0
plan           0
last_active    1
churned        0
dtype: int64


In [14]:
print("Exact duplicate rows:")

print(
    df[df.duplicated(keep=False)]
)

Exact duplicate rows:
  account_id    company            signup_ts country seats   mrr        plan  \
0     A-1001  Acme Corp  2023-01-05 09:00:00      US    25  2500  enterprise   
2     A-1001  Acme Corp  2023-01-05 09:00:00      US    25  2500  enterprise   

           last_active churned  
0  2024-03-01 00:00:00   False  
2  2024-03-01 00:00:00   False  


In [15]:
print("Repeated account IDs:")

repeated_accounts = df[
    df.duplicated(
        subset='account_id',
        keep=False
    )
].sort_values('account_id')

print(repeated_accounts)

Repeated account IDs:
   account_id    company            signup_ts country seats   mrr        plan  \
0      A-1001  Acme Corp  2023-01-05 09:00:00      US    25  2500  enterprise   
2      A-1001  Acme Corp  2023-01-05 09:00:00      US    25  2500  enterprise   
1      A-1002   Beta LLC  2023-02-14 00:00:00      US     5   450         pro   
12     A-1002   Beta LLC  2023-02-14 00:00:00      US     6   540         pro   

            last_active churned  
0   2024-03-01 00:00:00   False  
2   2024-03-01 00:00:00   False  
1   2024-03-02 00:00:00   False  
12  2024-03-06 00:00:00   False  


In [16]:
print("Country values:")
print(df['country'].value_counts(dropna=False))

Country values:
country
US               12
GB                1
United States     1
us                1
DE                1
CA                1
Name: count, dtype: int64


In [17]:
print("Plan values:")
print(df['plan'].value_counts(dropna=False))

Plan values:
plan
pro           11
enterprise     4
PRO            1
starter        1
Name: count, dtype: int64


In [21]:
# Convert seats and mrr from text to numbers

df['seats_num'] = pd.to_numeric(
    df['seats'],
    errors='coerce'
)

df['mrr_num'] = pd.to_numeric(
    df['mrr'],
    errors='coerce'
)

print(df[['account_id', 'seats', 'seats_num', 'mrr', 'mrr_num']])

   account_id    seats  seats_num   mrr  mrr_num
0      A-1001       25       25.0  2500     2500
1      A-1002        5        5.0   450      450
2      A-1001       25       25.0  2500     2500
3      A-1003        0        0.0     0        0
4      A-1004       12       12.0  1188     1188
5      A-1005        8        8.0   -99      -99
6      A-1006        3        3.0   270      270
7      A-1007       15       15.0  1800     1800
8      A-1008      NaN        NaN   540      540
9      A-1009       10       10.0   900      900
10     A-1010       20       20.0     0        0
11     A-1011        7        7.0   630      630
12     A-1002        6        6.0   540      540
13     A-1012  1000000  1000000.0    90       90
14     A-1013        4        4.0   360      360
15     A-1014        9        9.0   810      810
16     A-1015        6        6.0   540      540


In [22]:
print("Suspicious MRR values:")

print(
    df[
        (df['mrr_num'] < 0) |
        (df['mrr_num'] == 0)
    ][
        ['account_id', 'mrr', 'seats']
    ]
)

Suspicious MRR values:
   account_id  mrr seats
3      A-1003    0     0
5      A-1005  -99     8
10     A-1010    0    20


In [23]:
df['signup_parsed'] = pd.to_datetime(
    df['signup_ts'],
    errors='coerce'
)

df['last_active_parsed'] = pd.to_datetime(
    df['last_active'],
    errors='coerce'
)

print(
    df[
        [
            'account_id',
            'signup_ts',
            'signup_parsed',
            'last_active',
            'last_active_parsed'
        ]
    ]
)

   account_id            signup_ts       signup_parsed          last_active  \
0      A-1001  2023-01-05 09:00:00 2023-01-05 09:00:00  2024-03-01 00:00:00   
1      A-1002  2023-02-14 00:00:00 2023-02-14 00:00:00  2024-03-02 00:00:00   
2      A-1001  2023-01-05 09:00:00 2023-01-05 09:00:00  2024-03-01 00:00:00   
3      A-1003           1675209600                 NaT  2024-02-28 00:00:00   
4      A-1004  2023-03-20 00:00:00 2023-03-20 00:00:00  2024-03-01 00:00:00   
5      A-1005  2023-04-01 00:00:00 2023-04-01 00:00:00  2024-03-03 00:00:00   
6      A-1006  2023-05-11 00:00:00 2023-05-11 00:00:00  2024-03-02 00:00:00   
7      A-1007  2023-06-30 00:00:00 2023-06-30 00:00:00  2023-06-15 00:00:00   
8      A-1008  2023-07-22 00:00:00 2023-07-22 00:00:00  2024-03-04 00:00:00   
9      A-1009  2023-08-08 00:00:00 2023-08-08 00:00:00                  NaN   
10     A-1010  2023-09-01 00:00:00 2023-09-01 00:00:00  2024-03-01 00:00:00   
11     A-1011  2023-10-10 00:00:00 2023-10-10 00:00:

In [24]:
bad_date_order = (
    df['signup_parsed'].notna()
    &
    df['last_active_parsed'].notna()
    &
    (
        df['last_active_parsed']
        <
        df['signup_parsed']
    )
)

print("Last active before signup:")

print(
    df.loc[
        bad_date_order,
        [
            'account_id',
            'signup_ts',
            'last_active'
        ]
    ]
)

Last active before signup:
  account_id            signup_ts          last_active
7     A-1007  2023-06-30 00:00:00  2023-06-15 00:00:00


In [25]:
today = pd.Timestamp.today().normalize()

future_activity = (
    df['last_active_parsed'].notna()
    &
    (
        df['last_active_parsed'] > today
    )
)

print("Future last-active dates:")

print(
    df.loc[
        future_activity,
        [
            'account_id',
            'last_active'
        ]
    ]
)

Future last-active dates:
   account_id          last_active
14     A-1013  2099-01-01 00:00:00


In [26]:
print(
    df.loc[
        df['account_id'] == 'A-1003',
        [
            'account_id',
            'signup_ts',
            'signup_parsed'
        ]
    ]
)

  account_id   signup_ts signup_parsed
3     A-1003  1675209600           NaT


## Q5 Diagnosis

| Problem | Classification | Fix? | Reason |
|---|---|---|---|
| A-1001 appears twice with identical values | What is one row? | Yes | The two rows are exact duplicates and would double-count the same record. |
| A-1002 appears twice with different seats, MRR and last_active values | What is one row? | No automatic deletion | The repeated account appears to represent a real update rather than an exact duplicate. |
| A-1003 has signup_ts = 1675209600 | What should each column be? | Investigate/standardize | The value uses a different date representation from the other signup dates. |
| A-1004 has country = United States while other US records use US | What should each column be? | Yes, after confirming the standard | The country representation is inconsistent. |
| A-1005 has mrr = -99 | What is missing or lying? | Investigate/interpret | -99 is a code rather than a real negative MRR value, so its meaning should be confirmed before changing it. |
| A-1006 uses us, PRO and FALSE | What should each column be? | Yes | Country, plan and churned use inconsistent representations. |
| A-1007 has last_active before signup_ts | What is missing or lying? | Investigate | The dates violate the expected chronological relationship. |
| A-1008 has missing seats | What is missing or lying? | Investigate | Seats is a missing measure; missing does not necessarily mean zero. |
| A-1009 has missing last_active | What is missing or lying? | Investigate | The meaning of the missing activity date needs to be established before deleting or imputing it. |
| A-1010 has mrr = 0 | What is missing or lying? | Do not automatically fix | Zero can be a legitimate business value and should not automatically be treated as missing. |
| A-1012 has seats = 1000000 | What is missing or lying? | Investigate | The value creates an implausibly small MRR-per-seat value, but the data alone cannot establish which field is wrong. |
| A-1013 has last_active = 2099-01-01 | What is missing or lying? | Investigate/fix | The value is a valid date format but is impossible in the context of this dataset. |
| A-1014 has missing company | What is missing or lying? | Investigate | Company is a missing label and should be recovered from an authoritative source if possible. |

# Q6. The duplicate that isn't

     ### Prediction

I predict that `df.drop_duplicates()` will remove one of the two A-1001 rows because they are exact duplicates. I predict that both A-1002 rows will remain because their seats, MRR, and last_active values are different.

In [27]:
print("Rows before drop_duplicates():", len(df))

deduped = df.drop_duplicates()

print("Rows after drop_duplicates():", len(deduped))

Rows before drop_duplicates(): 17
Rows after drop_duplicates(): 16


In [28]:
print("A-1001 rows before:")
print(df[df['account_id'] == 'A-1001'])

print("\nA-1001 rows after:")
print(deduped[deduped['account_id'] == 'A-1001'])

print("\nA-1002 rows before:")
print(df[df['account_id'] == 'A-1002'])

print("\nA-1002 rows after:")
print(deduped[deduped['account_id'] == 'A-1002'])

A-1001 rows before:
  account_id    company            signup_ts country seats   mrr        plan  \
0     A-1001  Acme Corp  2023-01-05 09:00:00      US    25  2500  enterprise   
2     A-1001  Acme Corp  2023-01-05 09:00:00      US    25  2500  enterprise   

           last_active churned  seats_num  mrr_num       signup_parsed  \
0  2024-03-01 00:00:00   False       25.0     2500 2023-01-05 09:00:00   
2  2024-03-01 00:00:00   False       25.0     2500 2023-01-05 09:00:00   

  last_active_parsed  
0         2024-03-01  
2         2024-03-01  

A-1001 rows after:
  account_id    company            signup_ts country seats   mrr        plan  \
0     A-1001  Acme Corp  2023-01-05 09:00:00      US    25  2500  enterprise   

           last_active churned  seats_num  mrr_num       signup_parsed  \
0  2024-03-01 00:00:00   False       25.0     2500 2023-01-05 09:00:00   

  last_active_parsed  
0         2024-03-01  

A-1002 rows before:
   account_id   company            signup_ts count

### Actual result

`df.drop_duplicates()` reduced the dataset from 17 rows to 16 rows.

It removed one of the two A-1001 records because those rows are exact duplicates. Both A-1002 records remained because they are not identical: the seats, MRR, and last_active values differ.

### Why drop_duplicates() is the wrong tool for A-1002

`drop_duplicates()` compares complete rows. It does not decide whether two records with the same account_id represent duplicate data or legitimate historical updates.

A-1002 appears twice, but the second record changes from 5 seats and 450 MRR to 6 seats and 540 MRR, with a later last_active date. Therefore, I would not delete one of the A-1002 records without understanding what one row represents.

### Replacement approach

If the business requirement is to keep only the latest record for each account, I would first identify the correct date field and then use:

```python
df['last_active_dt'] = pd.to_datetime(
    df['last_active'],
    errors='coerce'
)

latest = (
    df.sort_values('last_active_dt')
      .drop_duplicates(
          subset='account_id',
          keep='last'
      )
)

# Q7 — CELL 2: Calculate MRR per seat

In [29]:
df['mrr_per_seat'] = (
    df['mrr_num'] / df['seats_num']
)

print(
    df[
        [
            'account_id',
            'seats_num',
            'mrr_num',
            'mrr_per_seat'
        ]
    ]
)

   account_id  seats_num  mrr_num  mrr_per_seat
0      A-1001       25.0     2500     100.00000
1      A-1002        5.0      450      90.00000
2      A-1001       25.0     2500     100.00000
3      A-1003        0.0        0           NaN
4      A-1004       12.0     1188      99.00000
5      A-1005        8.0      -99     -12.37500
6      A-1006        3.0      270      90.00000
7      A-1007       15.0     1800     120.00000
8      A-1008        NaN      540           NaN
9      A-1009       10.0      900      90.00000
10     A-1010       20.0        0       0.00000
11     A-1011        7.0      630      90.00000
12     A-1002        6.0      540      90.00000
13     A-1012  1000000.0       90       0.00009
14     A-1013        4.0      360      90.00000
15     A-1014        9.0      810      90.00000
16     A-1015        6.0      540      90.00000


### Q7 Diagnosis

The normal accounts cluster around approximately 90 MRR per seat.

**A-1003:** The row has 0 seats and 0 MRR, so MRR divided by seats is 0/0 and is undefined. I would not invent a per-seat value. The row may be legitimate if zero-seat, zero-MRR accounts are allowed.

**A-1005:** The value is -99/8 = -12.375 MRR per seat. The most likely explanation is that -99 is a special code rather than an actual negative MRR value. I would investigate the source-system definition and treat it accordingly rather than interpreting it as negative revenue.

**A-1008:** The seats value is missing, so MRR per seat cannot be calculated. I would investigate the missing seat value rather than assuming it is zero.

**A-1010:** The row has 20 seats and 0 MRR, producing 0 MRR per seat. This is unusual compared with the baseline but zero is a valid numerical value. I would leave it alone unless business evidence shows that an account with 20 seats cannot have zero MRR.

**A-1012:** The row has 1,000,000 seats and 90 MRR, producing 0.00009 MRR per seat. This is highly suspicious. There are two competing explanations: the seat count could be wrong, or the MRR could be wrong. The data alone cannot distinguish these explanations. An authoritative billing or account record would be needed.

**A-1007:** The value is 120 MRR per seat, which differs from the dominant 90 baseline but is not automatically evidence of bad data. I would not change it without additional business evidence.

# Q8. Parsing is not validating

In [30]:
print("Parsing failures:")

print(
    df.loc[
        df['signup_parsed'].isna(),
        [
            'account_id',
            'signup_ts'
        ]
    ]
)

Parsing failures:
  account_id   signup_ts
3     A-1003  1675209600


In [31]:
bad_order = (
    df['signup_parsed'].notna()
    &
    df['last_active_parsed'].notna()
    &
    (
        df['last_active_parsed']
        <
        df['signup_parsed']
    )
)

print("Last active before signup:")

print(
    df.loc[
        bad_order,
        [
            'account_id',
            'signup_ts',
            'last_active'
        ]
    ]
)

Last active before signup:
  account_id            signup_ts          last_active
7     A-1007  2023-06-30 00:00:00  2023-06-15 00:00:00


In [32]:
today = pd.Timestamp.today().normalize()

future_activity = (
    df['last_active_parsed'].notna()
    &
    (
        df['last_active_parsed'] > today
    )
)

print("Future last-active dates:")

print(
    df.loc[
        future_activity,
        [
            'account_id',
            'last_active'
        ]
    ]
)

Future last-active dates:
   account_id          last_active
14     A-1013  2099-01-01 00:00:00


### Q8 Explanation

A parsing problem occurs when the raw value cannot be interpreted using the expected data representation. A-1003 is an example because signup_ts contains 1675209600, which is not being interpreted as an ordinary date string by the parsing operation.

A validation problem occurs when a value can be successfully parsed but violates a business or logical rule. A-1007 has a valid date in last_active, but its last_active date is before its signup date. A-1013 is another example: 2099-01-01 is a valid date format, but it is an impossible future activity date in the context of this dataset.

# Q9. The same null, two fates

In [33]:
print("Missing company:")
print(
    df[
        df['company'].isna()
    ]
)

print("\nMissing seats:")
print(
    df[
        df['seats'].isna()
    ]
)

Missing company:
   account_id company            signup_ts country seats  mrr plan  \
15     A-1014     NaN  2024-01-15 00:00:00      US     9  810  pro   

            last_active churned  seats_num  mrr_num signup_parsed  \
15  2024-03-03 00:00:00   False        9.0      810    2024-01-15   

   last_active_parsed  mrr_per_seat  
15         2024-03-03          90.0  

Missing seats:
  account_id company            signup_ts country seats  mrr plan  \
8     A-1008   Helio  2023-07-22 00:00:00      US   NaN  540  pro   

           last_active churned  seats_num  mrr_num signup_parsed  \
8  2024-03-04 00:00:00   False        NaN      540    2023-07-22   

  last_active_parsed  mrr_per_seat  
8         2024-03-04           NaN  


### Q9 Explanation

A-1014 is missing a value in the label column `company`. A-1008 is missing a value in the measure column `seats`.

The same action, such as dropping the row, can have different consequences depending on the downstream analysis. If I were grouping MRR by company, A-1014 cannot be assigned to a company group, so excluding it from that particular analysis could be appropriate if the company cannot be recovered.

A-1008 is different because its MRR value is still present. If the downstream use is calculating total MRR, dropping A-1008 would unnecessarily remove valid revenue information. However, if the downstream use is calculating MRR per seat, A-1008 cannot be used until the missing seats value is resolved.

# Q10. Cleaning is a function of the question

### Decision 1: Removing the exact A-1001 duplicate

Earlier, I identified the repeated A-1001 row as an exact duplicate and would remove one copy when creating a clean account-level dataset. However, if the goal were to investigate source-system duplication or data-ingestion errors, I would keep both records because the duplicate itself would be useful evidence of the problem.

### Decision 2: Handling A-1008's missing seats

For an MRR-per-seat analysis, I would exclude A-1008 from the calculation until its missing seats value can be resolved because the ratio cannot be calculated. However, if the goal were to calculate total MRR, I would keep A-1008 because its MRR value is still valid and useful.

### Conclusion

"Clean" is not a permanent property of a dataset; it depends on the question the data is being used to answer. A record or value that should be removed or excluded for one analysis can be useful and should be retained for another analysis.

### Decision 1: Removing the exact A-1001 duplicate

Earlier, I identified the repeated A-1001 row as an exact duplicate and would remove one copy when creating a clean account-level dataset. However, if the goal were to investigate source-system duplication or data-ingestion errors, I would keep both records because the duplicate itself would be useful evidence of the problem.

### Decision 2: Handling A-1008's missing seats

For an MRR-per-seat analysis, I would exclude A-1008 from the calculation until its missing seats value can be resolved because the ratio cannot be calculated. However, if the goal were to calculate total MRR, I would keep A-1008 because its MRR value is still valid and useful.

### Conclusion

"Clean" is not a permanent property of a dataset; it depends on the question the data is being used to answer. A record or value that should be removed or excluded for one analysis can be useful and should be retained for another analysis.